ziomon/ziomon_mgr.8
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH ZIOMON_MGR 8 "Jul 2008" "s390-tools"

.SH NAME
ziomon_mgr \- Message queue server for the ziomon framework.

.SH SYNOPSIS
.B ziomon_mgr
[-h] [-v] [-V] [-e] [-f] [-l <size>] [-x <version>] -o <filename> -i <length> -Q <msgq_path> -q <msgq_id> -u <util_id> -r <ioerr_id> -b <blkiomon_id> -z <zfcpdd_id>

.SH DESCRIPTION
.B ziomon_mgr
starts a message queue where the clients of the ziomon framework can
send their data to. Any data received will be we written to a specified
file. Optionally, an upper limit for the file can be specified. If the
limit is exceeded, the oldest data will be aggregated into a separate
file to make room for the latest.

For consistent data, all clients should schedule their interval
lengths to the same duration. In general, clients should send their
data at the same time.

This command is not intended to be run on its own - rather use the ziomon command.

.SH OPTIONS
.TP
.BR "\-h" " or " "\-\-help"
Print help information, then exit.

.TP
.BR "\-v" " or " "\-\-version"
Print version information, then exit.

.TP
.BR "\-V" " or " "\-\-verbose"
Be verbose.

.TP
.BR "\-e" " or " "\-\-binary-offsets"
Lengths of the raw data structures.

.TP
.BR "\-f" " or " "\-\-force"
Force message queue creation in case one already exists.

.TP
.BR "\-o" " or " "\-\-output"
Basename of the file to write data to. Respective suffixes will be appended
for aggregated and regular data file names.

.TP
.BR "\-l" " or " "\-\-size-limit"
Upper limit of the output file in MB. This does not include the space for
the aggregated data file. However, the size of the aggregated data file
is usually negligible.

.TP
.BR "\-x" " or " "\-\-enforce-version"
Enforce specific file format for .log and .agg files. Currently supports
versions 2 (blkiomon version 0.2) and 3 (blkiomon version 0.3 or higher).

.TP
.BR "\-i" " or " "\-\-interval-length"
Expected elapsed time between messages sent by the clients in seconds.
The samples will be aggregated over this period. Defaults to the duration.
This should be a multiple of the sample length.

.TP
.BR "\-Q" " or " "\-\-msg-queue-name"
Name for the message queue to start.

.TP
.BR "\-q" " or " "\-\-msg-queue-id"
Id for the message queue to start. Must be an integer >0.

.TP
.BR "\-u" " or " "\-\-util-id"
Id of the utilization messages that ziomon_util will send.
Must be an integer >0.

.TP
.BR "\-r" " or " "\-\-ioerr-id"
Id of the ioerr messages that ziomon_util will send.
Must be an integer >0.

.TP
.BR "\-b" " or " "\-\-blkiomon-id"
Id of the messages that ziomon_util will send.
Must be an integer >0.

.TP
.BR "\-z" " or " "\-\-ziomon_zfcpdd-id"
Id of the messages that ziomon_zfcpdd will send.
Must be an integer >0.

.SH EXAMPLES
Start a message queue named /tmp/ziomon.23943 with id 95.
Utilization messages from ziomon_util will carry the id 17, ioerr messages 18,
messages from ziomon_zfcpdd
47 and messages from blkiomon 59. The interval between each set of messages will
be 20 seconds. The output files should base their names on 'trace_data':

ziomon_mgr -i 20 -Q /tmp/ziomon.23943 -q 95 -u 17 -r 18 -b 47 -z 59 -o trace_data

.SH "SEE ALSO"
.BR ziomon (8),
.BR ziomon_util (8),
.BR ziomon_zfcpdd (8)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------
* __v2.2x.x (2022-xx-xx)__

  For Linux kernel version: 5.xx

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.21.0 (2022-04-20)__

  For Linux kernel version: 5.17

  Add new tools / libraries:
  - libcpumf: Create library libcpumf for CPU Measurement functions

  Changes of existing tools:
  - chreipl-fcp-mpath: bundle a pre-cooked version of the manpage for build
                       environments without access to `pandoc`
  - dbginfo.sh: Add multipath info to map paths to FC addressing and prio group
  - dbginfo.sh: Collect config files of systemd-modules-load.service
  - dbginfo.sh: Sort list of environment variables for readability
  - dbginfo.sh: Replace "which" by builtin command "type"
  - dbginfo.sh: Rework script formatting (indents, order)
  - dbginfo.sh: Update sysfs collection (excludes, messages)
  - genprotimg: Add Protected Virtualization (PV) dump support
  - genprotimg: Remove DigiCert root CA pinning
  - lszcrypt: Add CEX8S support
  - zcryptctl: Add control domain handling
  - zcryptstats: Add CEX8 support
  - zipl: Allow optional entries that are left out when files are missing
  - zipl: make IPL sections defined with BLS to inherit a target field
  - zpcictl: Add option to trigger firmware reset

  Bug Fixes:
  - cpictl: Handle excessive kernel version numbers
  - dbginfo.sh: Collect all places where modprobe.d config files could exist
  - fdasd: Fix endless menu loop on EOF
  - zdump/dfi: Fix segfault due to double free
  - zdump: Fix /dev/mem reading
  - zpcictl: Fix race of SCLP reset and Linux recovery

* __v2.20.0 (2022-02-04)__

  For Linux kernel version: 5.16

  Add new tools / libraries:
  - Add EditorConfig configuration

  Changes of existing tools:
  - s390-tools switches to Fuse 3 as Fuse 2 is deprecated.
        Affected tools: cmsfs, hmcdrvfs, hsavmcore, zdsfs, zdump
  - chreipl-fcp-mpath: don't compress the manpage before installing it
  - cpictl: Report extended version information
  - genprotimg: Add extended kernel command line support
  - zdev: modify the lsblk output parser in lszdev
  - zipl: Add support for longer kernel command lines (now supports up to 64k length)

  Bug Fixes:
  - cpictl: Suppress messages for unwritable sysfs files
  - dbginfo.sh: Fix missing syslog for step create_package
  - lshwc: Fix CPU list parameter setup for device driver
  - zdev: Check for errors when removing a devtype setting
  - zdev: Fix path resolution for multi-mount point file systems

* __v2.19.0 (2021-11-10)__

  For Linux kernel version: 5.15

  Add new tools / libraries:
  - chreipl-fcp-mpath: New toolset that uses multipath information to change
      the configured FCP re-IPL path on detecting issues with the current path

  Changes of existing tools:
  - dbginfo.sh: Add retry timeout and remove possible blocking "blockdev --report"
  - dbginfo.sh: Collect config- and debug-data for chreipl-fcp-mpath
  - hsci: Add support for multiple MAC addresses

  Bug Fixes:
  - lshwc: Fix compile error for gcc <8.1
  - zdump: Various clean-ups and fixes
  - ziomon: Correct throughput calculation in ziorep_printers
  - zipl: Fix segmentation fault when setting stage3_parms

* __v2.18.0 (2021-10-01)__

  For Linux kernel version: 5.14

  Add new tools:
  - scripts: Add tool for parsing sclp s390dbf logs
  - zdev: Add udev rule helper tool
  - zipl-editenv: Add tool to operate with zIPL environment installed in the boot record

  Changes of existing tools:
  - Makefile: Fix order of build of libraries for parallel builds
  - dbginfo.sh: Add collection in area of timedate, coredump and --check option
  - dbginfo.sh: Add exception on dump2tar for /sys/kernel/mm/page_idle/bitmap
  - dbginfo.sh: Cleanup of outdated sections and general code rework
  - dbginfo.sh: Collect zipl boot menu entries from boot loader specification
  - lszcrypt: Add support for vfio-ap status field
  - lszcrypt: Improved output for deconfig cards and queues
  - lszfcp: Add linkdown case to host marker of extended output
  - zdev: Add auto-config for PCI and crypto devices
  - zdump: Introduce multi-level message logging
  - zipl: Add support for environment block interpretation
  - zkey-cryptsetup: Support LUKS2 volumes with integrity support enabled

  Bug Fixes:
  - hsavmcore: Avoid recompilation of overlay during install step
  - libkmipclient: Fix parsing of hex values for XML and JSON encoding
  - vmur/vmur.cpp: Fix error handling on transfer failure
  - zdump: Lots of smaller fixes across the board

* __v2.17.0 (2021-07-07)__

  For Linux kernel version: 5.12 / 5.13

  Add new tools / libraries:
  - hsavmcore: New utility to make the dump process with kdump more efficient
  - libkmipclient: Add KMIP client shared library
  - libseckey: Add a secure key library
  - lshwc: New tool to extract and list complete counter sets

  Changes of existing tools:
  - genprotimg: Add '--(enable|disable)-pckmo' options
  - genprotimg: Add OpenSSL 3.0 support
  - genprotimg: Change plaintext control flags defaults so PCKMO functions are allowed
  - libutil: Introduce multi-level message logging (util_log)
  - libutil: Introduce util_arch module
  - udev/dasd: Change DASD udev-rule to set none scheduler
  - zdsfs: Add transparent codepage conversion
  - zkey: Add support for KMIP-based key management systems

  Bug Fixes:
  - ttyrun-getty: Avoid conflicts with serial-getty@
  - dbginfo: add /proc/kallsyms - refresh zVM, lscpu - fix WORKARCHIVE handling
  - dbginfo: add KVM data collection for server and guest - fix lszdev
  - genprotimg: Add missing return values in error paths
  - zkey: Fix conversion of CCA DATA keys to CCA CIPHER keys
  - znetconf: avoid conflict with "chzdev -e"

* __v2.16.0 (2021-02-19)__

  For Linux kernel version: 5.10 / 5.11

  Add new tool:
  - hsci: New tool to manage HSCI (HiperSockets Converged Interfaces)

  Changes of existing tools:

ziomon/ziomon_mgr.c
/*
 * FCP adapter trace utility
 *
 * Central message collection tool
 *
 * Copyright IBM Corp. 2008, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <linux/types.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ipc.h>
#include <sys/msg.h>
#include <sys/msg.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <time.h>
#include <unistd.h>

#include "lib/zt_common.h"

#include "ziomon_dacc.h"
#include "ziomon_msg_tools.h"
#include "ziomon_tools.h"
#include "ziomon_util.h"
#include "ziomon_zfcpdd.h"
#include "blkiomon.h"



const char *toolname = "ziomon_mgr";
int verbose=0;
static int keep_running = 1;


struct options {
	char   		       *msg_q_path;
	int			msg_q_id;
	int			msg_q;
	long			msg_id_utilization;
	long			msg_id_ioerr;
	long			msg_id_blkiomon;
	long			msg_id_zfcpdd;
	int			estimate;
	int			interval_length;
	int			force;
	long                    version;
	char   		       *outfile_name;
	char   		       *outfile_name_agg;
	FILE   		       *outfile;
	FILE		       *outfile_agg;
	struct aggr_data	agg_data;
	long			size_limit;
	short			wrapped;
	struct file_header	f_hdr;
};


static void init_opts(struct options *opts)
{
	opts->msg_q_path = NULL;
	opts->msg_q_id = -1;
	opts->msg_q = -1;
	opts->msg_id_blkiomon = LONG_MIN;
	opts->msg_id_utilization = LONG_MIN;
	opts->msg_id_ioerr = LONG_MIN;
	opts->msg_id_zfcpdd = LONG_MIN;
	opts->outfile_name = NULL;
	opts->outfile_name_agg = NULL;
	opts->outfile = NULL;
	opts->outfile_agg = NULL;
	opts->size_limit = LONG_MAX;
	opts->wrapped = 0;
	opts->interval_length = -1;
	opts->force = 0;
	opts->estimate = 0;
	opts->version = 3;
}


static void deinit_opts(struct options *opts)
{
	if (opts->msg_q >= 0) {
		verbose_msg("shutting down message queue\n");
		if (msgctl(opts->msg_q, IPC_RMID, 0) == -1)
			fprintf(stderr, "%s: Error encountered"
				" while shutting down message queue: %s\n",
				toolname, strerror(errno));
	}
	if (opts->outfile)
		fclose(opts->outfile);
	free(opts->outfile_name);
	free(opts->outfile_name_agg);
	if (opts->outfile_agg) {
		fclose(opts->outfile_agg);
		discard_aggr_data_struct(&opts->agg_data);
	}
}


static const char help_text[] =
  "Usage: ziomon_mgr [-h] [-v] [-V] [-e] [-f] [-l <size>] [-x <version>]"
  " -o <filename> -i <length>\n"
  "                  -Q <msgq-path> -q <msgq-id> -u <util-id> -r <ioerr-id>\n"
  "                  -b <blkiomon-id> -z <ziomon_zfcpdd-id>\n"
  "Start the message server for the ziomon framework.\n"
  "\n"
  "-h, --help              Print usage information and exit.\n"
  "-v, --version           Print version information and exit.\n"
  "-V, --verbose           Be verbose.\n"
  "-e, --binary-offsets    Print a list of binary data structure"
                         " sizes and exit.\n"
  "-f, --force             Force message queue creation.\n"
  "-i, --interval-length   Specify interval length in seconds.\n"
  "-Q, --msg-queue-name    Specify the message queue path name.\n"
  "-q, --msg-queue-id      Specify the message queue id.\n"
  "-u, --util-id           Specify the id for utilization messages from\n"
  "                        ziomon_util.\n"
  "-r, --ioerr-id          Specify the id for ioerr messages from"
			   " ziomon_util.\n"
  "-b, --blkiomon-id       Specify the id for messages from blkiomon.\n"
  "-z, --ziomon-zfcpdd-id  Specify the id for messages from ziomon_zfcpdd.\n"
  "-o, --output            Specify the name of the output file(s).\n"
  "-l, --size-limit        Maximum size of data collected in MB.\n"
  "-x, --enforce-version   Enforce specific version for .log and .agg files.\n";

static void print_help(void)
{
        fprintf(stdout, "%s", help_text);
}


static void print_version(void)
{
        fprintf(stdout, "%s: ziomon manager, version %s\n"
               "Copyright IBM Corp. 2008, 2017\n",
               toolname, RELEASE_STRING);
}


static int setup_msg_q(struct options *opts)
{
	key_t util_q;
	int flags;

	util_q = ftok(opts->msg_q_path, opts->msg_q_id);

	if (util_q == -1) {
		fprintf(stderr, "%s: Cannot create"
			" message queue with path %s and id %d: %s\n",
			toolname, opts->msg_q_path, opts->msg_q_id, strerror(errno));
		return -1;
	}

	verbose_msg("message queue key is %d\n", util_q);

	flags = IPC_CREAT | S_IRWXU;
	if (!opts->force)
		flags |= IPC_EXCL;
	opts->msg_q = msgget(util_q, flags);
	if (opts->msg_q < 0) {
		fprintf(stderr, "%s: Could not open message queue"
			": %s\n", toolname, strerror(errno));
		if (!opts->force)
			fprintf(stderr, "%s: Retry using the 'force'"
				" option\n", toolname);
		return -1;
	}
	verbose_msg("message queue id is %d\n", opts->msg_q);

	return 0;
}


static int add_to_aggregated(struct message **msgs, int num_msgs,
			     struct options *opts)
{
	int i;

	if (!opts->outfile_agg) {
		opts->outfile_agg = fopen(opts->outfile_name_agg, "w+");
		if (!opts->outfile_agg) {
			fprintf(stderr, "%s: Could not open file"
				" for aggregated data: %s\n", toolname,
				strerror(errno));
			return -1;
		}
		init_aggr_data_struct(&opts->agg_data);
	}

	/* aggregate data */
	for (i = 0; i < num_msgs; ++i) {
		if (add_to_agg(&opts->agg_data, msgs[i], &opts->f_hdr))
			return -1;
		discard_msg(msgs[i]);
		free(msgs[i]);
	}
	free(msgs);

	/* write back to file */
	conv_aggr_data_msg_data_to_BE(&opts->agg_data);
	i = write_aggr_file(opts->outfile_agg, &opts->agg_data);
	conv_aggr_data_msg_data_from_BE(&opts->agg_data);

	return i;
}

static int compare_msg_ids(const void *a, const void *b)
{
	return (*(long*)b - *(long*)a);
}

static int check_msg_ids(struct options *opts)
{
	int i;
	long msg_ids[4];

	msg_ids[0] = opts->msg_id_utilization;
	msg_ids[1] = opts->msg_id_ioerr;
	msg_ids[2] = opts->msg_id_blkiomon;
	msg_ids[3] = opts->msg_id_zfcpdd;

	qsort(msg_ids, 4, sizeof(long), compare_msg_ids);

	for (i = 0; i < 4; ++i) {
		if (msg_ids[i] <= 0) {
			fprintf(stderr, "%s: ids for messages must be"
				" greater than 0\n", toolname);
			return -1;
		}
		if (i != 0 && msg_ids[i] == msg_ids[i-1]) {
			fprintf(stderr, "%s: ids for messages must be"
				" different from another\n", toolname);
			return -1;
		}
	}

	return 0;
}


void print_bin_struct_sizes(void)
{
	fprintf(stdout, "%ld %ld %ld %ld %ld %ld\n",
		(unsigned long int)sizeof(struct utilization_data),
		(unsigned long int)sizeof(struct adapter_utilization),
		(unsigned long int)sizeof(struct ioerr_data),
		(unsigned long int)sizeof(struct ioerr_cnt),
		(unsigned long int)sizeof(struct zfcpdd_dstat),
		(unsigned long int)sizeof(struct blkiomon_stat));
}


static int convert_long_optarg(long *tgt, char *option_str)
{
	if (!optarg) {
		fprintf(stderr, "%s: Error: Argument missing to"
			" option %s\n", toolname, option_str);
		return -1;
	}
	*tgt = strtol(optarg, NULL, 0);
	if (errno) {
		fprintf(stderr, "%s: Error converting %s to"
			" long: %s\n", toolname, optarg, strerror(errno));
		return -1;
	}

	return 0;
}


static int parse_params(int argc, char **argv, struct options *opts)
{
	int c;
	int error=0;
	int index;
        static struct option long_options[] = {
                { "version",         no_argument,       NULL, 'v'},
		{ "help",            no_argument,       NULL, 'h'},
		{ "binary-offsets",  no_argument,       NULL, 'e'},
                { "interval-length", required_argument, NULL, 'i'},
		{ "msg-queue-name",  required_argument, NULL, 'Q'},
		{ "msg-queue-id",    required_argument, NULL, 'q'},
		{ "util-id",         required_argument, NULL, 'u'},
		{ "ioerr-id",        required_argument, NULL, 'r'},
		{ "blkiomon-id",     required_argument, NULL, 'b'},
		{ "ziomon-zfcp-id",  required_argument, NULL, 'z'},
		{ "verbose",         no_argument,       NULL, 'V'},
		{ "size-limit",      required_argument, NULL, 'l'},
		{ "enforce-version", required_argument, NULL, 'x'},
		{ "output",          required_argument, NULL, 'o'},
		{ "force",           no_argument,       NULL, 'f'},
                { 0,                 0,                 0,     0 }
	};

	if (argc <= 1) {
		print_help();
		exit(EXIT_FAILURE);
	}

	while ((c = getopt_long(argc, argv, "r:Q:q:u:b:z:i:l:o:x:Vhfev",
				long_options, &index)) != EOF) {
		switch (c) {
		case 'V':
			verbose++;
			break;
		case 'Q':
			if (!optarg) {
				fprintf(stderr, "%s: Error:"
					" Argument missing to option '-Q'\n",
					toolname);
				return -1;
			}
			opts->msg_q_path = optarg;
			break;
		case 'q':
			if (!optarg) {
				fprintf(stderr, "%s: Error:"
					" Argument missing to option '-q'\n",
					toolname);
				return -1;
			}
			opts->msg_q_id = atoi(optarg);
			if (opts->msg_q_id < 0) {
				fprintf(stderr, "%s: Error:"
					" Parameter to option '-q' must be"
					" greater than 0\n", toolname);
				return -1;
			}
			break;
		case 'u':
			if (convert_long_optarg(&opts->msg_id_utilization,
						"'-u'"))
				return -1;
			break;
		case 'r':
			if (convert_long_optarg(&opts->msg_id_ioerr,
						"'-r"))
				return -1;
			break;
		case 'b':
			if (convert_long_optarg(&opts->msg_id_blkiomon,
						"'-b'"))
				return -1;
			break;
		case 'z':
			if (convert_long_optarg(&opts->msg_id_zfcpdd,
						"'-z'"))
				return -1;
			break;
		case 'h':
			print_help();
			exit(EXIT_SUCCESS);
		case 'e':
			print_bin_struct_sizes();
			return 1;
		case 'f':
			opts->force = 1;
			break;
		case 'i':
			if (!optarg) {
				fprintf(stderr, "%s: Argument missing to"
					" option '-i'\n", toolname);
				return -1;
			}
			opts->interval_length = atoi(optarg);
			if (opts->interval_length <= 0) {
				fprintf(stderr, "%s: Interval"
					" length must be >0\n", toolname);
				return -1;
			}
			break;
		case 'o':
			if (!optarg) {
				fprintf(stderr, "%s: Argument missing to"
					" option '-o'\n", toolname);
				return -1;
			}
			if (opts->outfile_name) {
				fprintf(stderr, "%s: Multiple instance of "
					" option '-o'\n", toolname);
				return -2;
			}
			opts->outfile_name = malloc(strlen(optarg)
					+ strlen(DACC_FILE_EXT_LOG) + 1);
			if (!opts->outfile_name) {
				fprintf(stderr, "%s: Memory allocation "
					"error\n", toolname);
				return -3;
			}
			opts->outfile_name_agg = malloc(strlen(optarg)
					+ strlen(DACC_FILE_EXT_AGG) + 1);
			sprintf(opts->outfile_name, "%s" DACC_FILE_EXT_LOG,
				optarg);
			sprintf(opts->outfile_name_agg, "%s" DACC_FILE_EXT_AGG,
				optarg);
			break;
		case 'l':
			if (!optarg) {
				fprintf(stderr, "%s: Argument missing to"
					" option '-o'\n", toolname);
				return -1;
			}
			opts->size_limit = strtol(optarg, NULL, 0);
			if (errno) {
				fprintf(stderr, "%s: Error during conversion:"
					" %s\n", toolname, strerror(errno));
				return -1;
			}
			if (opts->size_limit < 1) {
				fprintf(stderr, "%s: Size limit must be >=1.\n",
					toolname);
				return -1;
			}
			opts->size_limit *= 1024*1024;
			break;
		case 'x':
			if (!optarg) {
				fprintf(stderr, "%s: Argument missing to"
					" option '-x'\n", toolname);
				return -1;
			}
			opts->version = strtol(optarg, NULL, 0);
			if (errno) {
				fprintf(stderr, "%s: Error during conversion:"
					" %s\n", toolname, strerror(errno));
				return -1;
			}
			if (opts->version < 2 || opts->version > 3) {
				fprintf(stderr, "%s: Enforced version can only be"
					"2 or 3.\n", toolname);
				return -1;
			}
			break;
		case 'v':
			print_version();
			exit(EXIT_SUCCESS);
		default:
			fprintf(stderr, "Try '%s --help' for"
				" more information.\n", toolname);
			return -1;
		}
	}

	if (!opts->msg_q_path) {
		fprintf(stderr, "%s: No message queue path specified\n",
				toolname);
		error++;
	}
	if (opts->msg_q_id < 0) {
		fprintf(stderr, "%s: No message queue"
			" id specified\n", toolname);
		error++;
	}
	if (opts->msg_id_blkiomon == LONG_MIN) {
		fprintf(stderr, "%s: No id for blkiomon"
			" messages specified\n", toolname);
		error++;
	}
	if (opts->msg_id_utilization == LONG_MIN) {
		fprintf(stderr, "%s: No id for"
			" utilizaton messages specified\n", toolname);
		error++;
	}
	if (opts->msg_id_ioerr == LONG_MIN) {
		fprintf(stderr, "%s: No id for"
			" ioerr messages specified\n", toolname);
		error++;
	}
	if (opts->msg_id_zfcpdd == LONG_MIN) {
		fprintf(stderr, "%s: No id for"
			" ziomon_zfcpdd messages specified\n", toolname);
		error++;
	}
	if (opts->interval_length < 0) {
		fprintf(stderr, "%s: No interval length"
			" specified\n", toolname);
		error++;
	}
	if (!opts->outfile_name) {
		fprintf(stderr, "%s: No filename for"
			" output specified\n", toolname);
		error++;
	}
	if (error)
		return -1;

	if (check_msg_ids(opts))
		return -1;

	opts->outfile = fopen(opts->outfile_name, "w+");
	if (!opts->outfile) {
		fprintf(stderr, "%s: Could not open output"
			" file: %s\n", toolname, strerror(errno));
		return -1;
	}

	if (setup_msg_q(opts))
		return -1;

	verbose_msg("interval length      : %d\n", opts->interval_length);
	verbose_msg("force                : %d\n", opts->force);
	verbose_msg("message queue path   : %s\n", opts->msg_q_path);
	verbose_msg("message queue id     : %d\n", opts->msg_q_id);
	verbose_msg("message queue        : %d\n", opts->msg_q);
	verbose_msg("msg id utilization   : %ld\n", opts->msg_id_utilization);
	verbose_msg("msg id ioerr         : %ld\n", opts->msg_id_ioerr);
	verbose_msg("msg id blkiomon      : %ld\n", opts->msg_id_blkiomon);
	verbose_msg("msg id ziomon_zfcpdd : %ld\n", opts->msg_id_zfcpdd);
	verbose_msg("outfile name         : %s\n", opts->outfile_name);
	if (opts->size_limit == LONG_MAX)
		verbose_msg("size limit           : no limit\n");
	else
		verbose_msg("size limit           : %ld Bytes\n", opts->size_limit);

	return 0;
}


static void void_handler(int sig)
{
	verbose_msg("interrupted by signal %u\n", sig);
	keep_running = 0;
}


static void print_timestamp(struct tm *my_tm, const char *type, __u32 length,
			    struct timeval *t)
{
	verbose_msg("%02d:%02d:%02d.%06ld: received %s"
		    " message of %u bytes\n", my_tm->tm_hour,
		    my_tm->tm_min, my_tm->tm_sec, t->tv_usec, type, length);
}


static int handle_msg(struct message *msg, struct options *opts)
{
	struct message **msgs;
	int count;
	int rc = 0;
	struct timeval t;
	struct tm *my_tm = NULL;

	if (verbose) {
		gettimeofday(&t, NULL);
		my_tm = localtime(&t.tv_sec);
	}

	if ((long)msg->type == opts->msg_id_utilization)
		print_timestamp(my_tm, "utilization", msg->length, &t);
	else if ((long)msg->type == opts->msg_id_ioerr)
		print_timestamp(my_tm, "ioerr", msg->length, &t);
	else if ((long)msg->type == opts->msg_id_blkiomon)
		print_timestamp(my_tm, "blkiomon", msg->length, &t);
	else if ((long)msg->type == opts->msg_id_zfcpdd)
		print_timestamp(my_tm, "zfcpdd", msg->length, &t);
	else {
		fprintf(stderr, "%s: Received message of "
				"unrecognized type %d, length %d bytes, "
				"discarding\n", toolname, msg->type, msg->length);
		return 1;
	}

	if (add_msg(opts->outfile, msg, &opts->f_hdr, &msgs, &count)) {
		fprintf(stderr, "%s: Error while writing"
			" message\n", toolname);
		rc = -1;
	} else
		verbose_msg("message written\n");
	if (count) {
		if (add_to_aggregated(msgs, count, opts)) {
			fprintf(stderr, "%s: Failed to aggregate"
				" %d messages\n", toolname, count);
			rc = -1;
		}
	}

	return rc;
}


int main(int argc, char **argv)
{
	int rc = 0;
	struct options opts;
	int len;
	int data_sz = 1024;
	long *data = malloc(data_sz + sizeof(long));
	int tmperr;
	struct message msg;

	verbose = 0;

	signal(SIGALRM, void_handler);
	signal(SIGINT,  void_handler);
	signal(SIGTERM, void_handler);
	signal(SIGQUIT, void_handler);

	init_opts(&opts);

	if (parse_params(argc, argv, &opts))
		goto out;

	/* first write to file, init section for cumulatives */
	verbose_msg("init file...\n");
	opts.f_hdr.msgid_utilization = opts.msg_id_utilization;
	opts.f_hdr.msgid_ioerr = opts.msg_id_ioerr;
	opts.f_hdr.msgid_blkiomon = opts.msg_id_blkiomon;
	opts.f_hdr.msgid_zfcpdd = opts.msg_id_zfcpdd;
	opts.f_hdr.size_limit = opts.size_limit;
	opts.f_hdr.interval_length = opts.interval_length;
	if (init_file(opts.outfile, &opts.f_hdr, opts.version))
		goto out;

	verbose_msg("wait for messages...\n");
	do {
		len = msgrcv(opts.msg_q, data, data_sz, 0, 0);
		if (!keep_running)
			break;
		if (len < 0) {
			tmperr = errno;
			if (tmperr == E2BIG) {
				data_sz *= 2;
				data = realloc(data, data_sz + sizeof(long));
				verbose_msg("message buffer too small,"
					    " increasing to %d\n", data_sz);
				continue;
			}
			fprintf(stderr, "%s: Error receiving"
				" message: %s\n", toolname, strerror(errno));
			verbose_msg("msgrcv() returned error %d\n", tmperr);
			break;
		}
		msg.length = len;
		msg.data = data + 1;
		msg.type = *data;
		handle_msg(&msg, &opts);

	} while (keep_running);

out:
	deinit_opts(&opts);
	free(data);

	return rc;
}



